### PR TITLE
Reduce mutability in Phase related code

### PR DESCRIPTION
**Pull request by yurloc**
The goal of this PR is to make it easier to understand and reason about the Phases code. Declaring fields `final` and eliminating the corresponding setters reduces the effort of searching for the places from where the field can be changed. With `final`, the intention is clear: the value is determined during object construction and then never changes.

The Builder pattern is taken from Item 2, Effective Java. In this case it solves the problem with optional assertion flags. The Builder has a named setter method for each optional argument, so instead of something like
```java
XyPhase phase;
if (environment.isLevelA() && environment.isLevelB()) {
    phase = new XyPhase(requiredArgs, true, true);
}
if (environment.isLevelA()) {
    phase = new XyPhase(requiredArgs, true);
}
// etc.
```
and multiple telescoping constructors for `XyPhase`, we have something where the client code is almost the same as before but immutability is achieved once the object is built:
```java
// Initilialize builder with mandatory arguments.
Builder builder = new XyPhase.Builder(requiredArgs);
// Conditionally enable any combination of optional assertion flags.
if (environment.isLevelA()) {
    builder.setAssertStepScoreFromScratch(true);
}
if (environment.isLevelB()) {
    builder.setAssertExpectedStepScore(true);
}
// Once the object is built, even the optional arguments become final.
XyPhase phase = builder.build();
```

<details>
<summary>
How to replicate CI configuration locally?
</summary>

We do "simple" maven builds, they are just basically maven commands, but just because we have multiple repositories related between them and one change could affect several of those projects by multiple pull requests, we use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.

[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is not only a github-action tool but a CLI one, so in case you posted multiple pull requests related with this change you can easily reproduce the same build by executing it locally. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] tests</b>
* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
* for a <b>specific Native LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] native</b>
</details>

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/HeuristicConfigPolicy.java [hunk 1]
```diff
@@ -144,23 +113,19 @@ public boolean isInitializedChainedValueFilterEnabled() {
...
+                .withLogIndentation(logIndentation);
     }
 
-    public HeuristicConfigPolicy<Solution_> createFilteredPhaseConfigPolicy() {
-        HeuristicConfigPolicy<Solution_> heuristicConfigPolicy = createPhaseConfigPolicy();
-        heuristicConfigPolicy.reinitializeVariableFilterEnabled = true;
-        heuristicConfigPolicy.initializedChainedValueFilterEnabled = true;
-        return heuristicConfigPolicy;
+    public HeuristicConfigPolicy<Solution_> createPhaseConfigPolicy() {
+        return buildClone().build();
```
Comment (yurloc) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/HeuristicConfigPolicy.java:122]: Or maybe `clonerBuilder().build()`?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java [hunk 2]
```diff
@@ -269,4 +243,35 @@ public void solvingEnded(SolverScope<Solution_> solverScope) {
         decider.solvingEnded(solverScope);
     }
 
+    public static class Builder<Solution_> extends AbstractPhase.Builder<Solution_> {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java:246]: Do we need a builder pattern?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java [hunk 3]
```diff
@@ -269,4 +243,35 @@ public void solvingEnded(SolverScope<Solution_> solverScope) {
...
+            this.assertWorkingSolutionScoreFromScratch = assertWorkingSolutionScoreFromScratch;
+        }
+
+        public void setAssertExpectedWorkingSolutionScore(boolean assertExpectedWorkingSolutionScore) {
+            this.assertExpectedWorkingSolutionScore = assertExpectedWorkingSolutionScore;
+        }
+
+        @Override
+        public DefaultExhaustiveSearchPhase<Solution_> build() {
+            return new DefaultExhaustiveSearchPhase<>(this);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java:274]: Typically, in a builder pattern, the builder itself is not given to the thing it builds IIRC.

**Comment by ge0ffrey** [comment]
> ```java
> XyPhase phase;
> if (environment.isLevelA() && environment.isLevelB()) {
>     phase = new XyPhase(requiredArgs, true, true);
> }
> if (environment.isLevelA()) {
>     phase = new XyPhase(requiredArgs, true);
> }
> // etc.
> ```
Personally, I prefer this before code over the after code. What do the other think?

**Comment by ge0ffrey** [comment]
+1 to try to make more fields final of the phases (not the scopes), where it doesn't increase complexity.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/HeuristicConfigPolicy.java (hunk 1 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/HeuristicConfigPolicy.java:122]: Sounds much better, as the `buildClone` method is not creating the clone itself.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java (hunk 3 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java:274]: An alternative to passing a builder instance is creating a constructor that accepts all possible options as parameters. In that case, this `build` method becomes yet another place where the options need to be passed to the constructor as arguments. For that reason, I prefer the way @yurloc proposes.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java (hunk 2 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java:246]: It depends on if all the properties are mandatory. If they are not, removing setters leads to a combinatorial explosion of constructors.

**Review by rsynek — APPROVED** [review]

**Review by triceo — APPROVED** [review]
LGTM. The endless mutability of core has been bugging me for a while.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java (hunk 3 above)
Comment (yurloc) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/exhaustivesearch/DefaultExhaustiveSearchPhase.java:274]: Exactly as Radek says. Passing the builder instance avoids repetition. I've seen this approach in multiple sources, including the Effective Java book.

**Comment by yurloc** [comment]
> Do we need a builder pattern? Why not just use a simple constructor? KISS principle.

I agree the Builder code is a boilerplate but I think it does pull its weight:
- It's pretty straightforward and it's nicely separated at the end of its enclosing class so it doesn't interrupt reading it.
- It fits the current state where all the affected classes have some optional parameters that are being conditionally set based on the `Environment`. It preserves the existing API that makes the client code read nicely. Optional parameters are being set using named methods (`builder.setAssertExpectedStepScore(true)`) vs. being passed as an anonymous value in the constructor (`new DefaultLocalSearchPhase(..., true)`) while allowing them to stay immutable once the object is constructed.

**Comment by triceo** [comment]
2 approvals, no rejections, merging.

**Pull request merged**